I tried running the Linux CUDA x64 CI several times. It seems there is a reproducible segmentation fault that needs to be resolved.

```
Fatal Python error: Segmentation fault

Current thread 0x0000756a1ac39680 (most recent call first):
  File "/ort_genai_src/test/python/test_onnxruntime_genai_api.py", line 153 in test_greedy_search
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/python.py", line 167 in pytest_pyfunc_call
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_callers.py", line 121 in _multicall
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_hooks.py", line 512 in __call__
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/python.py", line 1707 in runtest
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/runner.py", line 184 in pytest_runtest_call
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_callers.py", line 121 in _multicall
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_hooks.py", line 512 in __call__
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/runner.py", line 250 in <lambda>
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/runner.py", line 361 in from_call
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/runner.py", line 249 in call_and_report
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/runner.py", line 139 in runtestprotocol
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/runner.py", line 118 in pytest_runtest_protocol
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_callers.py", line 121 in _multicall
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_hooks.py", line 512 in __call__
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/main.py", line 408 in pytest_runtestloop
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_callers.py", line 121 in _multicall
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_hooks.py", line 512 in __call__
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/main.py", line 384 in _main
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/main.py", line 330 in wrap_session
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/main.py", line 377 in pytest_cmdline_main
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_callers.py", line 121 in _multicall
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pluggy/_hooks.py", line 512 in __call__
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/config/__init__.py", line 229 in _main
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/_pytest/config/__init__.py", line 253 in _console_main
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pytest/__main__.py", line 9 in <module>
  File "<frozen runpy>", line 88 in _run_code
  File "<frozen runpy>", line 198 in _run_module_as_main

Extension modules: numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, google._upb._message, torch._C, torch._C._dynamo.autograd_compiler, torch._C._dynamo.eval_frame, torch._C._dynamo.guards, torch._C._dynamo.utils, torch._C._fft, torch._C._linalg, torch._C._nested, torch._C._nn, torch._C._sparse, torch._C._special, charset_normalizer.md, charset_normalizer.cd, requests.packages.charset_normalizer.md, requests.packages.chardet.md, requests.packages.charset_normalizer.cd, requests.packages.chardet.cd, yaml._yaml, regex._regex, markupsafe._speedups (total: 34)
Traceback (most recent call last):
  File "/ort_genai_src/test/python/test_onnxruntime_genai.py", line 118, in <module>
    sys.exit(main())
             ^^^^^^
  File "/ort_genai_src/test/python/test_onnxruntime_genai.py", line 109, in main
    run_onnxruntime_genai_api_tests(os.path.abspath(args.cwd), log, os.path.abspath(args.test_models))
  File "/ort_genai_src/test/python/test_onnxruntime_genai.py", line 35, in run_onnxruntime_genai_api_tests
    run_subprocess(command, cwd=cwd, log=log).check_returncode()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/ort_genai_src/test/python/_test_utils.py", line 121, in run_subprocess
    completed_process = subprocess.run(
                        ^^^^^^^^^^^^^^^
  File "/opt/python/cp312-cp312/lib/python3.12/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/opt/python/cp312-cp312/bin/python3.12', '-m', 'pytest', '-sv', 'test_onnxruntime_genai_api.py', 'builder', 'models', '--test_models', '/ort_genai_src/test/models']' died with <Signals.SIGSEGV: 11>.
test_onnxruntime_genai_api.py::test_greedy_search[relative_model_path1] 
Error: Process completed with exit code 1.
```

Once that is resolved, we can re-run the CIs and get this PR merged in.

--- a/src/cpu/interface.cpp
+++ b/src/cpu/interface.cpp
@@ -59,6 +59,10 @@ struct CpuInterface : DeviceInterface {
     return *ort_allocator_;
   }
 
+  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
+    throw std::runtime_error("GetMemoryInfo for CPU should not be used.");
+  }
+
   std::shared_ptr<DeviceBuffer> AllocateBase(size_t size) override {
     return std::make_shared<CpuMemory>(size);
   }
@@ -166,10 +170,6 @@ struct CpuInterface : DeviceInterface {
   std::unique_ptr<Search> CreateBeam(const GeneratorParams& params) override { return std::make_unique<BeamSearch_Cpu>(params); }
 
   void Synchronize() override {}  // Nothing to do as CPU is always in sync with itself
-
-  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
-    throw std::runtime_error("GetMemoryInfo for CPU should not be used.");
-  }
 };
 
 std::unique_ptr<DeviceInterface> CreateCpuInterface() {

--- a/src/cpu/interface.cpp
+++ b/src/cpu/interface.cpp
@@ -59,10 +59,6 @@ struct CpuInterface : DeviceInterface {
     return *ort_allocator_;
   }
 
-  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
-    throw std::runtime_error("GetMemoryInfo for CPU should not be used.");
-  }
-
   std::shared_ptr<DeviceBuffer> AllocateBase(size_t size) override {
     return std::make_shared<CpuMemory>(size);
   }
@@ -170,6 +166,10 @@ struct CpuInterface : DeviceInterface {
   std::unique_ptr<Search> CreateBeam(const GeneratorParams& params) override { return std::make_unique<BeamSearch_Cpu>(params); }
 
   void Synchronize() override {}  // Nothing to do as CPU is always in sync with itself
+
+  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
+    throw std::runtime_error("GetMemoryInfo for CPU should not be used.");
+  }
 };
 
 std::unique_ptr<DeviceInterface> CreateCpuInterface() {

--- a/src/cuda/interface.cpp
+++ b/src/cuda/interface.cpp
@@ -90,6 +90,13 @@ struct CudaInterfaceImplBase : DeviceInterface {
     return *ort_allocator_;
   }
 
+  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
+    return OrtMemoryInfo::Create("Cuda",
+                                 OrtAllocatorType::OrtDeviceAllocator,
+                                 0,
+                                 OrtMemType::OrtMemTypeDefault);
+  }
+
   std::shared_ptr<DeviceBuffer> AllocateBase(size_t size) override {
     return std::make_shared<GpuMemory>(size);
   }
@@ -177,13 +184,6 @@ struct CudaInterfaceImplBase : DeviceInterface {
 
   void GetAvailableMemory(size_t& free_bytes, size_t& total_bytes) override {
     cudaMemGetInfo(&free_bytes, &total_bytes);
-  }
-
-  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
-    return OrtMemoryInfo::Create("Cuda",
-                                 OrtAllocatorType::OrtDeviceAllocator,
-                                 0,
-                                 OrtMemType::OrtMemTypeDefault);
   }
 };
 

--- a/src/cuda/interface.cpp
+++ b/src/cuda/interface.cpp
@@ -90,13 +90,6 @@ struct CudaInterfaceImplBase : DeviceInterface {
     return *ort_allocator_;
   }
 
-  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
-    return OrtMemoryInfo::Create("Cuda",
-                                 OrtAllocatorType::OrtDeviceAllocator,
-                                 0,
-                                 OrtMemType::OrtMemTypeDefault);
-  }
-
   std::shared_ptr<DeviceBuffer> AllocateBase(size_t size) override {
     return std::make_shared<GpuMemory>(size);
   }
@@ -184,6 +177,13 @@ struct CudaInterfaceImplBase : DeviceInterface {
 
   void GetAvailableMemory(size_t& free_bytes, size_t& total_bytes) override {
     cudaMemGetInfo(&free_bytes, &total_bytes);
+  }
+
+  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
+    return OrtMemoryInfo::Create("Cuda",
+                                 OrtAllocatorType::OrtDeviceAllocator,
+                                 0,
+                                 OrtMemType::OrtMemTypeDefault);
   }
 };
 

--- a/src/cuda/interface.cpp
+++ b/src/cuda/interface.cpp
@@ -81,7 +81,6 @@ struct CudaInterfaceImplBase : DeviceInterface {
   }
 
   void InitOrt(const OrtApi& api, Ort::Allocator& allocator) override {
-    Ort::api = &api;
     assert(!ort_allocator_);
     ort_allocator_ = &allocator;
   }
@@ -246,8 +245,11 @@ void operator delete(void* p, size_t /*size*/) noexcept {
 #endif
 
 extern "C" {
-Generators::DeviceInterface* GetInterface(GenaiInterface* p_genai, const char* deviceType) {
+Generators::DeviceInterface* GetInterface(GenaiInterface* p_genai, const char* deviceType, const OrtApi* ort_api) {
   Generators::gp_genai = p_genai;
+  // Ensure Ort::api is initialized in this shared library (onnxruntime-genai-cuda add-on) immediately. Delaying the
+  // initialization to CudaInterfaceImplBase::InitOrt would be inadequate, as GetMemoryInfo runs before InitOrt.
+  Ort::api = ort_api;
   if (strcasecmp(deviceType, "NvTensorRtRtx") == 0) {
     Generators::g_cuda_device = std::make_unique<Generators::NvTensorRtRtxInterfaceImpl>();
   } else {

--- a/src/dml/interface.cpp
+++ b/src/dml/interface.cpp
@@ -137,13 +137,6 @@ struct InterfaceImpl : DeviceInterface {
     return *ort_allocator_;
   }
 
-  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
-    return OrtMemoryInfo::Create("DML",
-                                 OrtAllocatorType::OrtDeviceAllocator,
-                                 0,
-                                 OrtMemType::OrtMemTypeDefault);
-  }
-
   std::shared_ptr<DeviceBuffer> AllocateBase(size_t size) override {
     return std::make_shared<GpuMemory>(size);
   }
@@ -213,6 +206,13 @@ struct InterfaceImpl : DeviceInterface {
 #endif
 
   void Synchronize() override {}
+
+  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
+    return OrtMemoryInfo::Create("DML",
+                                 OrtAllocatorType::OrtDeviceAllocator,
+                                 0,
+                                 OrtMemType::OrtMemTypeDefault);
+  }
 };
 
 }  // namespace Dml

--- a/src/dml/interface.cpp
+++ b/src/dml/interface.cpp
@@ -137,6 +137,13 @@ struct InterfaceImpl : DeviceInterface {
     return *ort_allocator_;
   }
 
+  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
+    return OrtMemoryInfo::Create("DML",
+                                 OrtAllocatorType::OrtDeviceAllocator,
+                                 0,
+                                 OrtMemType::OrtMemTypeDefault);
+  }
+
   std::shared_ptr<DeviceBuffer> AllocateBase(size_t size) override {
     return std::make_shared<GpuMemory>(size);
   }
@@ -206,13 +213,6 @@ struct InterfaceImpl : DeviceInterface {
 #endif
 
   void Synchronize() override {}
-
-  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
-    return OrtMemoryInfo::Create("DML",
-                                 OrtAllocatorType::OrtDeviceAllocator,
-                                 0,
-                                 OrtMemType::OrtMemTypeDefault);
-  }
 };
 
 }  // namespace Dml

--- a/src/generators.cpp
+++ b/src/generators.cpp
@@ -288,9 +288,9 @@ DeviceInterface* OrtGlobals::LoadCudaInterface(DeviceType type) {
     if (!*cuda_library_)
       throw std::runtime_error("Shared library load failure (see first error)");
 
-    Generators::DeviceInterface* GetInterface(GenaiInterface * p_genai, const char* deviceType);
+    Generators::DeviceInterface* GetInterface(GenaiInterface * p_genai, const char* deviceType, const OrtApi* ort_api);
     return reinterpret_cast<decltype(&GetInterface)>(
-        cuda_library_->GetSymbol("GetInterface"))(&g_genai, to_string(type).c_str());
+        cuda_library_->GetSymbol("GetInterface"))(&g_genai, to_string(type).c_str(), Ort::api);
   } catch (const std::exception& e) {
     throw std::runtime_error("Cuda interface not available: " + std::string(e.what()));
   }

--- a/src/openvino/interface.cpp
+++ b/src/openvino/interface.cpp
@@ -25,10 +25,6 @@ struct InterfaceImpl : DeviceInterface {
     return GetDeviceInterface(DeviceType::CPU)->GetAllocator();
   }
 
-  virtual std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
-    throw std::runtime_error("GetMemoryInfo for OpenVINO should not be used. Expected to use CPU interface instead.");
-  }
-
   std::shared_ptr<DeviceBuffer> AllocateBase(size_t size) override {
     return GetDeviceInterface(DeviceType::CPU)->AllocateBase(size);
   }
@@ -41,6 +37,10 @@ struct InterfaceImpl : DeviceInterface {
   std::unique_ptr<Search> CreateBeam(const GeneratorParams& params) override { return std::make_unique<BeamSearch_Cpu>(params); }
 
   void Synchronize() override {}  // Nothing to do
+
+  virtual std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
+    throw std::runtime_error("GetMemoryInfo for OpenVINO should not be used. Expected to use CPU interface instead.");
+  }
 };
 
 }  // namespace OpenVINO

--- a/src/openvino/interface.cpp
+++ b/src/openvino/interface.cpp
@@ -25,6 +25,10 @@ struct InterfaceImpl : DeviceInterface {
     return GetDeviceInterface(DeviceType::CPU)->GetAllocator();
   }
 
+  virtual std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
+    throw std::runtime_error("GetMemoryInfo for OpenVINO should not be used. Expected to use CPU interface instead.");
+  }
+
   std::shared_ptr<DeviceBuffer> AllocateBase(size_t size) override {
     return GetDeviceInterface(DeviceType::CPU)->AllocateBase(size);
   }
@@ -37,10 +41,6 @@ struct InterfaceImpl : DeviceInterface {
   std::unique_ptr<Search> CreateBeam(const GeneratorParams& params) override { return std::make_unique<BeamSearch_Cpu>(params); }
 
   void Synchronize() override {}  // Nothing to do
-
-  virtual std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
-    throw std::runtime_error("GetMemoryInfo for OpenVINO should not be used. Expected to use CPU interface instead.");
-  }
 };
 
 }  // namespace OpenVINO

--- a/src/qnn/interface.cpp
+++ b/src/qnn/interface.cpp
@@ -44,6 +44,12 @@ struct QnnMemory final : DeviceBuffer {
 };
 
 struct QnnInterfaceBase : DeviceInterface {
+  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
+    // Note: "QnnHtpShared" allocator is the correct name for both HTP and GPU backends. Eventually, the plan is to
+    // migrate to "QnnShared".
+    return OrtMemoryInfo::Create("QnnHtpShared", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
+  }
+
   void InitOrt(const OrtApi& /*api*/, Ort::Allocator& allocator) override {
     assert(!ort_allocator_);
     ort_allocator_ = &allocator;
@@ -65,12 +71,6 @@ struct QnnInterfaceBase : DeviceInterface {
   std::unique_ptr<Search> CreateBeam(const GeneratorParams& params) override { return std::make_unique<BeamSearch_Cpu>(params); }
 
   void Synchronize() override {}  // Nothing to do
-
-  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
-    // Note: "QnnHtpShared" allocator is the correct name for both HTP and GPU backends. Eventually, the plan is to
-    // migrate to "QnnShared".
-    return OrtMemoryInfo::Create("QnnHtpShared", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
-  }
 };
 
 struct HtpInterfaceImpl : QnnInterfaceBase {

--- a/src/qnn/interface.cpp
+++ b/src/qnn/interface.cpp
@@ -44,12 +44,6 @@ struct QnnMemory final : DeviceBuffer {
 };
 
 struct QnnInterfaceBase : DeviceInterface {
-  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
-    // Note: "QnnHtpShared" allocator is the correct name for both HTP and GPU backends. Eventually, the plan is to
-    // migrate to "QnnShared".
-    return OrtMemoryInfo::Create("QnnHtpShared", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
-  }
-
   void InitOrt(const OrtApi& /*api*/, Ort::Allocator& allocator) override {
     assert(!ort_allocator_);
     ort_allocator_ = &allocator;
@@ -71,6 +65,12 @@ struct QnnInterfaceBase : DeviceInterface {
   std::unique_ptr<Search> CreateBeam(const GeneratorParams& params) override { return std::make_unique<BeamSearch_Cpu>(params); }
 
   void Synchronize() override {}  // Nothing to do
+
+  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
+    // Note: "QnnHtpShared" allocator is the correct name for both HTP and GPU backends. Eventually, the plan is to
+    // migrate to "QnnShared".
+    return OrtMemoryInfo::Create("QnnHtpShared", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
+  }
 };
 
 struct HtpInterfaceImpl : QnnInterfaceBase {

--- a/src/smartptrs.h
+++ b/src/smartptrs.h
@@ -108,7 +108,6 @@ struct DeviceInterface {
   virtual DeviceType GetType() const = 0;
   virtual void InitOrt(const OrtApi& api, Ort::Allocator& allocator) = 0;
   virtual Ort::Allocator& GetAllocator() = 0;
-  virtual std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const = 0;
 
   template <typename T>
   DeviceSpan<T> Allocate(size_t count) { return DeviceSpan<T>(AllocateBase(sizeof(T) * count)); }
@@ -150,6 +149,8 @@ struct DeviceInterface {
     assert(false);
     return nullptr;
   }  // Temporary until we fully factor out providers
+
+  virtual std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const = 0;
 };
 
 // A shared_ptr based type that we expose through our C API should inherit from this type.

--- a/src/smartptrs.h
+++ b/src/smartptrs.h
@@ -108,6 +108,7 @@ struct DeviceInterface {
   virtual DeviceType GetType() const = 0;
   virtual void InitOrt(const OrtApi& api, Ort::Allocator& allocator) = 0;
   virtual Ort::Allocator& GetAllocator() = 0;
+  virtual std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const = 0;
 
   template <typename T>
   DeviceSpan<T> Allocate(size_t count) { return DeviceSpan<T>(AllocateBase(sizeof(T) * count)); }
@@ -149,8 +150,6 @@ struct DeviceInterface {
     assert(false);
     return nullptr;
   }  // Temporary until we fully factor out providers
-
-  virtual std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const = 0;
 };
 
 // A shared_ptr based type that we expose through our C API should inherit from this type.

--- a/src/webgpu/interface.cpp
+++ b/src/webgpu/interface.cpp
@@ -182,6 +182,24 @@ struct InterfaceImpl : DeviceInterface {
     return *ort_allocator_;
   }
 
+  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
+    try {
+      return OrtMemoryInfo::Create("WebGPU_Buf", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
+    } catch (const Ort::Exception& e) {
+      // WebGPU memory type name changed from "WebGPU_Buffer" to "WebGPU_Buf" in ORT 1.24.3.
+      // Try the old name before giving up.
+      try {
+        return OrtMemoryInfo::Create("WebGPU_Buffer", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
+      } catch (const Ort::Exception& fallback_e) {
+        throw std::runtime_error(
+            "Failed to create memory info for WebGPU. "
+            "Primary name 'WebGPU_Buf' error: " +
+            std::string(e.what()) +
+            "; fallback 'WebGPU_Buffer' error: " + std::string(fallback_e.what()));
+      }
+    }
+  }
+
   std::shared_ptr<DeviceBuffer> AllocateBase(size_t size) override {
     return std::make_shared<WebGPUMemory>(size, ort_allocator_, ort_memory_info_);
   }
@@ -314,24 +332,6 @@ struct InterfaceImpl : DeviceInterface {
     for (const auto& opt : user_options->options) {
       if (std::find(kWebGpuGlobalOptions.begin(), kWebGpuGlobalOptions.end(), opt.first) != kWebGpuGlobalOptions.end()) {
         init_options.options.emplace_back(opt);
-      }
-    }
-  }
-
-  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
-    try {
-      return OrtMemoryInfo::Create("WebGPU_Buf", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
-    } catch (const Ort::Exception& e) {
-      // WebGPU memory type name changed from "WebGPU_Buffer" to "WebGPU_Buf" in ORT 1.24.3.
-      // Try the old name before giving up.
-      try {
-        return OrtMemoryInfo::Create("WebGPU_Buffer", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
-      } catch (const Ort::Exception& fallback_e) {
-        throw std::runtime_error(
-            "Failed to create memory info for WebGPU. "
-            "Primary name 'WebGPU_Buf' error: " +
-            std::string(e.what()) +
-            "; fallback 'WebGPU_Buffer' error: " + std::string(fallback_e.what()));
       }
     }
   }

--- a/src/webgpu/interface.cpp
+++ b/src/webgpu/interface.cpp
@@ -182,24 +182,6 @@ struct InterfaceImpl : DeviceInterface {
     return *ort_allocator_;
   }
 
-  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
-    try {
-      return OrtMemoryInfo::Create("WebGPU_Buf", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
-    } catch (const Ort::Exception& e) {
-      // WebGPU memory type name changed from "WebGPU_Buffer" to "WebGPU_Buf" in ORT 1.24.3.
-      // Try the old name before giving up.
-      try {
-        return OrtMemoryInfo::Create("WebGPU_Buffer", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
-      } catch (const Ort::Exception& fallback_e) {
-        throw std::runtime_error(
-            "Failed to create memory info for WebGPU. "
-            "Primary name 'WebGPU_Buf' error: " +
-            std::string(e.what()) +
-            "; fallback 'WebGPU_Buffer' error: " + std::string(fallback_e.what()));
-      }
-    }
-  }
-
   std::shared_ptr<DeviceBuffer> AllocateBase(size_t size) override {
     return std::make_shared<WebGPUMemory>(size, ort_allocator_, ort_memory_info_);
   }
@@ -332,6 +314,24 @@ struct InterfaceImpl : DeviceInterface {
     for (const auto& opt : user_options->options) {
       if (std::find(kWebGpuGlobalOptions.begin(), kWebGpuGlobalOptions.end(), opt.first) != kWebGpuGlobalOptions.end()) {
         init_options.options.emplace_back(opt);
+      }
+    }
+  }
+
+  std::unique_ptr<OrtMemoryInfo> GetMemoryInfo() const override {
+    try {
+      return OrtMemoryInfo::Create("WebGPU_Buf", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
+    } catch (const Ort::Exception& e) {
+      // WebGPU memory type name changed from "WebGPU_Buffer" to "WebGPU_Buf" in ORT 1.24.3.
+      // Try the old name before giving up.
+      try {
+        return OrtMemoryInfo::Create("WebGPU_Buffer", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
+      } catch (const Ort::Exception& fallback_e) {
+        throw std::runtime_error(
+            "Failed to create memory info for WebGPU. "
+            "Primary name 'WebGPU_Buf' error: " +
+            std::string(e.what()) +
+            "; fallback 'WebGPU_Buffer' error: " + std::string(fallback_e.what()));
       }
     }
   }

